### PR TITLE
Feat new spec formats

### PR DESCRIPTION
**Pull request by eldadfux**
## What does this PR do?

Instead if relying on an HTTP endpoint to generate our specs in realtime for our latest version, we can use the static files for showing specs for different supported versions.

I have added a new CLI command to change the way we generate the specs:
```bash
docker-compose exec appwrite specs --version=0.12.x
docker-compose exec appwrite specs --version=latest
docker-compose exec appwrite specs --mode=tests #for the sdk-generator mock structure
```

This will help to solve the issue where some of the paths in the homepage are missing, and help reduce the dependencies between appwrite/appwrite and appwrite.io.
https://appwrite.io/specs/open-api3?platform=client
https://appwrite.io/specs/open-api3?platform=server
https://appwrite.io/specs/swagger2?platform=client
https://appwrite.io/specs/swagger2?platform=server

## Test Plan

Added new tests to tests the generated spec files. Currently the tests are only validating the latest versions.

## Related PRs and Issues

#2527

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

Yes

### CODE REVIEW

File: app/tasks/specs.php [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+        $response = new Response(new HttpResponse());
+        $tests = ($mode === 'tests');
+        
+        App::setResource('request', function() {
+            return new Request;
+        });
+
+        App::setResource('response', function() use ($response) {
+            return $response;
+        });
```
Comment (TorstenDittmann) [inline app/tasks/specs.php:34]:
```suggestion
        App::setResource('request', fn() => new Request);
        App::setResource('response', fn() => $response);
```

we can be fancy here now :-)

File: src/Appwrite/Specification/Format/Swagger2.php [hunk 2]
```diff
@@ -460,17 +467,19 @@ public function parse(): array
...
                             } else {
                                 $items = [
-                                    'oneOf' => \array_map(function($type) {
+                                    'x-oneOf' => \array_map(function($type) {
                                         return ['$ref' => '#/definitions/'.$type];
                                     }, $rule['type'])
                                 ];
                             }
+
+                            // $items['$ref'] = '#/definitions/document';
```
Comment (TorstenDittmann) [inline src/Appwrite/Specification/Format/Swagger2.php:482]: leftover comment

File: tests/e2e/General/HTTPTest.php [hunk 3]
```diff
@@ -128,26 +128,41 @@ public function testSpecOpenAPI3()
...
+            ], json_decode(file_get_contents($directory.$file), true));
+
+            $response['body'] = json_decode($response['body'], true);
+
+            $this->assertEquals(200, $response['headers']['status-code']);
+            
+                        if(!empty($response['body'])){
+                            var_dump($directory.$file);
+                            var_dump($response['body']);
+                        }
```
Comment (TorstenDittmann) [inline tests/e2e/General/HTTPTest.php:163]: leftovers 👍🏻

Comment (christyjacob4) [inline tests/e2e/General/HTTPTest.php:163]: Leftovers I believe?

**Review by christyjacob4 — APPROVED** [review]

File: app/tasks/specs.php [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+        
+                        if ($platform !== APP_PLATFORM_CONSOLE && !\in_array($platforms[$platform], $sdkPlatofrms)) {
+                            continue;
+                        }
+        
+                        $routes[] = $route;
+                        $modelLabel = $route->getLabel('sdk.response.model', 'none');
+                        $model = \is_array($modelLabel) ? \array_map(function($m) use($response) {
+                            return $response->getModel($m);
+                        }, $modelLabel) : $response->getModel($modelLabel);
```
Comment (TorstenDittmann) [inline app/tasks/specs.php:189]: `$model` is never used 🤔

File: src/Appwrite/Specification/Format/OpenAPI3.php [hunk 5]
```diff
@@ -396,7 +396,14 @@ public function parse(): array
         foreach ($this->models as $model) {
             foreach ($model->getRules() as $rule) {
                 if (!in_array($rule['type'], ['string', 'integer', 'boolean', 'json', 'float'])) {
-                    $usedModels[] = $rule['type'];
+                    if(\is_array($rule['type'])) {
+                        foreach ($rule['type'] as $key => $value) {
```
Comment (TorstenDittmann) [inline src/Appwrite/Specification/Format/OpenAPI3.php:400]:
```suggestion
                        foreach ($rule['type'] as $value) {
```

we never use the `$key` here

File: src/Appwrite/Specification/Format/Swagger2.php [hunk 6]
```diff
@@ -393,7 +393,14 @@ public function parse(): array
         foreach ($this->models as $model) {
             foreach ($model->getRules() as $rule) {
                 if (!in_array($rule['type'], ['string', 'integer', 'boolean', 'json', 'float'])) {
-                    $usedModels[] = $rule['type'];
+                    if(\is_array($rule['type'])) {
+                        foreach ($rule['type'] as $key => $value) {
```
Comment (TorstenDittmann) [inline src/Appwrite/Specification/Format/Swagger2.php:397]:
```suggestion
                        foreach ($rule['type'] as $value) {
```

File: app/cli.php [hunk 7]
```diff
@@ -1,6 +1,6 @@
 <?php
-
 require_once __DIR__.'/init.php';
+require_once __DIR__.'/controllers/general.php';
```
Comment (TorstenDittmann) [inline app/cli.php:3]: This should be in `specs.php` since it adds like 2-4 seconds delay for every CLI command 👍🏻

**Review by TorstenDittmann — CHANGES_REQUESTED** [review]

File: app/cli.php (hunk 7 above)
Comment (TorstenDittmann) [reply app/cli.php:3]: nvm, looks like my docker was just a bit funky and slow

**Review by TorstenDittmann — APPROVED** [review]

**Pull request merged**